incipyt/project/meta_variables.py

```diff
@@ -0,0 +1,87 @@
+import dataclasses
+import logging
+import sys
+from datetime import date
+from typing import Any
+
+from incipyt._internal.utils import strtobool
+
+logger = logging.getLogger(__name__)
+
+# the public index where every variable metadata is saved
+variables = {}
+
+
+@dataclasses.dataclass
+class Variable:
+    """Metadata dataclass for environment variables.
+
+    If a specific `environ` variable hasn't been set, its metadata is used to constraint
+    how the prompt is done. If no metadata has been explicitly register, the prompt
+    behaves as a metadata with default field has been register.
+
+    Field `default: Any = ""`:
+
+    .. code-block::
+
+        Variable("VARIABLE_NAME", default="something")
+
+    The `default` value is used as a suggestion for the prompt.
+
+    Field `prompt: bool = True`:
+
+    .. code-block::
+
+        Variable("VARIABLE_NAME", prompt=False)
+
+    If False, the prompt is bypassed and the `default` value is used.
+
+    Field `required: bool = False`:
+
+    .. code-block::
+
+        EnvMetadata("VARIABLE_NAME", required=True)
+
+    .. code-block::
+
+        Variable("VARIABLE_NAME", type=int)
+
+    The `type` function is used to parse the value when it comes from a string,
+    e.g. to parse the value from the os environment variables.
+
+    When there is no `default` suggestion, if no value is given when prompt asks for it
+    `None` is saved in `environ` and all :class:`incipyt._internal.templates.Formattable`
+    values will be purge from template dictionary before commiting. When `True`, an empty
+    string is saved in `environ`. Note that if the `default` suggestion is not empty,
+    `required` has no effect as the prompt result will never be an empty string.
+    """
+
+    name: str
+    default: Any = ""
```

Comment:
We should make the default read-only, so people update the environ, not the defaults